os/hal/ports/AT32/LLD/DMAv1/DMAv1/at32_dma.c

```diff
@@ -0,0 +1,796 @@
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+/**
+ * @file    DMAv1/at32_dma.c
+ * @brief   DMA helper driver code.
+ *
+ * @addtogroup AT32_DMA
+ * @{
+ */
+
+#include "hal.h"
+
+/* The following macro is only defined if some driver requiring DMA services
+   has been enabled.*/
+#if defined(AT32_DMA_REQUIRED) || defined(__DOXYGEN__)
+
+/*===========================================================================*/
+/* Driver local definitions.                                                 */
+/*===========================================================================*/
+
+/**
+ * @brief   Mask of the DMA1 streams in @p dma_streams_mask.
+ */
+#define AT32_DMA1_STREAMS_MASK     ((1U << AT32_DMA1_NUM_CHANNELS) - 1U)
+
+/**
+ * @brief   Mask of the DMA2 streams in @p dma_streams_mask.
+ */
+#define AT32_DMA2_STREAMS_MASK     (((1U << AT32_DMA2_NUM_CHANNELS) -     \
+                                      1U) << AT32_DMA1_NUM_CHANNELS)
+
+#if AT32_DMA_SUPPORTS_CSELR == TRUE
+
+#if defined(DMA1_CSELR)
+#define __DMA1_CSELR                &DMA1_CSELR->CSELR
+#else
+#define __DMA1_CSELR                &DMA1->CSELR
+#endif
+
+#if defined(DMA2_CSELR)
+#define __DMA2_CSELR                &DMA2_CSELR->CSELR
+#else
+#define __DMA2_CSELR                &DMA2->CSELR
+#endif
+
+#define DMA1_CH1_VARIANT            __DMA1_CSELR
+#define DMA1_CH2_VARIANT            __DMA1_CSELR
+#define DMA1_CH3_VARIANT            __DMA1_CSELR
+#define DMA1_CH4_VARIANT            __DMA1_CSELR
+#define DMA1_CH5_VARIANT            __DMA1_CSELR
+#define DMA1_CH6_VARIANT            __DMA1_CSELR
+#define DMA1_CH7_VARIANT            __DMA1_CSELR
+#define DMA1_CH8_VARIANT            __DMA1_CSELR
+#define DMA2_CH1_VARIANT            __DMA2_CSELR
+#define DMA2_CH2_VARIANT            __DMA2_CSELR
+#define DMA2_CH3_VARIANT            __DMA2_CSELR
+#define DMA2_CH4_VARIANT            __DMA2_CSELR
+#define DMA2_CH5_VARIANT            __DMA2_CSELR
+#define DMA2_CH6_VARIANT            __DMA2_CSELR
+#define DMA2_CH7_VARIANT            __DMA2_CSELR
+#define DMA2_CH8_VARIANT            __DMA2_CSELR
+
+#else 
+
+#define DMA1_CH1_VARIANT            0
+#define DMA1_CH2_VARIANT            0
+#define DMA1_CH3_VARIANT            0
+#define DMA1_CH4_VARIANT            0
+#define DMA1_CH5_VARIANT            0
+#define DMA1_CH6_VARIANT            0
+#define DMA1_CH7_VARIANT            0
+#define DMA2_CH1_VARIANT            0
+#define DMA2_CH2_VARIANT            0
+#define DMA2_CH3_VARIANT            0
+#define DMA2_CH4_VARIANT            0
+#define DMA2_CH5_VARIANT            0
+#define DMA2_CH6_VARIANT            0
+#define DMA2_CH7_VARIANT            0
+
+#endif
```

Comment:
Since DMA_CSELR is only available for STM32F0/L0, so those are unused for AT32F415.
```suggestion
#define DMA1_CH1_VARIANT            0
#define DMA1_CH2_VARIANT            0
#define DMA1_CH3_VARIANT            0
#define DMA1_CH4_VARIANT            0
#define DMA1_CH5_VARIANT            0
#define DMA1_CH6_VARIANT            0
#define DMA1_CH7_VARIANT            0
#define DMA2_CH1_VARIANT            0
#define DMA2_CH2_VARIANT            0
#define DMA2_CH3_VARIANT            0
#define DMA2_CH4_VARIANT            0
#define DMA2_CH5_VARIANT            0
#define DMA2_CH6_VARIANT            0
#define DMA2_CH7_VARIANT            0
```